kmod/man/rmmod.8
'\" t
.\"     Title: rmmod
.\"    Author: Jon Masters <jcm@jonmasters.org>
.\" Generator: DocBook XSL Stylesheets v1.79.1 <http://docbook.sf.net/>
.\"      Date: 02/23/2017
.\"    Manual: rmmod
.\"    Source: kmod
.\"  Language: English
.\"
.TH "RMMOD" "8" "02/23/2017" "kmod" "rmmod"
.\" -----------------------------------------------------------------
.\" * Define some portability stuff
.\" -----------------------------------------------------------------
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.\" http://bugs.debian.org/507673
.\" http://lists.gnu.org/archive/html/groff/2009-02/msg00013.html
.\" ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\" -----------------------------------------------------------------
.\" * set default formatting
.\" -----------------------------------------------------------------
.\" disable hyphenation
.nh
.\" disable justification (adjust text to left margin only)
.ad l
.\" -----------------------------------------------------------------
.\" * MAIN CONTENT STARTS HERE *
.\" -----------------------------------------------------------------
.SH "NAME"
rmmod \- Simple program to remove a module from the Linux Kernel
.SH "SYNOPSIS"
.HP \w'\fBrmmod\fR\ 'u
\fBrmmod\fR [\fB\-f\fR] [\fB\-s\fR] [\fB\-v\fR] [\fImodulename\fR]
.SH "DESCRIPTION"
.PP
\fBrmmod\fR
is a trivial program to remove a module (when module unloading support is provided) from the kernel\&. Most users will want to use
\fBmodprobe\fR(8)
with the
\fB\-r\fR
option instead\&.
.SH "OPTIONS"
.PP
\fB\-v\fR, \fB\-\-verbose\fR
.RS 4
Print messages about what the program is doing\&. Usually
\fBrmmod\fR
prints messages only if something goes wrong\&.
.RE
.PP
\fB\-f\fR, \fB\-\-force\fR
.RS 4
This option can be extremely dangerous: it has no effect unless CONFIG_MODULE_FORCE_UNLOAD was set when the kernel was compiled\&. With this option, you can remove modules which are being used, or which are not designed to be removed, or have been marked as unsafe (see
\fBlsmod\fR(8))\&.
.RE
.PP
\fB\-s\fR, \fB\-\-syslog\fR
.RS 4
Send errors to syslog instead of standard error\&.
.RE
.PP
\fB\-V\fR \fB\-\-version\fR
.RS 4
Show version of program and exit\&.
.RE
.SH "COPYRIGHT"
.PP
This manual page originally Copyright 2002, Rusty Russell, IBM Corporation\&. Maintained by Jon Masters and others\&.
.SH "SEE ALSO"
.PP
\fBmodprobe\fR(8),
\fBinsmod\fR(8),
\fBlsmod\fR(8),
\fBmodinfo\fR(8)
.SH "AUTHORS"
.PP
\fBJon Masters\fR <\&jcm@jonmasters\&.org\&>
.RS 4
Developer
.RE
.PP
\fBLucas De Marchi\fR <\&lucas\&.de\&.marchi@gmail\&.com\&>
.RS 4
Developer
.RE


dbus/NEWS
dbus 1.13.18 (2020-07-02)
=========================

The “carnivorous border” release.

Maybe security fixes:

• On Unix, avoid a use-after-free if two usernames have the same
  numeric uid. In older versions this could lead to a crash (denial of
  service) or other undefined behaviour, possibly including incorrect
  authorization decisions if <policy group=...> is used.
  Like Unix filesystems, D-Bus' model of identity cannot distinguish
  between users of different names with the same numeric uid, so this
  configuration is not advisable on systems where D-Bus will be used.
  Thanks to Daniel Onaca.
  (dbus#305, dbus!166; Simon McVittie)

Other fixes:

• On Solaris and its derivatives, if a cmsg header is truncated, ensure
  that we do not overrun the buffer used for fd-passing, even if the
  kernel tells us to.
  (dbus#304, dbus!165; Andy Fiddaman)

• When built with CMake, use GNUInstallDirs' special-cases for prefixes
  /, /usr and /opt/*
  (dbus!155, Ralf Habacker)

• When built with CMake on Linux, allow systemd-specific features to be
  enabled, for feature parity with Autotools
  (dbus!155, Ralf Habacker)

• When built with CMake, install the same example files as with Autotools
  (dbus!155, Ralf Habacker)

• Correct the doc-comment for DBUS_ERROR_SPAWN_NO_MEMORY
  (dbus!163, Marc-André Lureau)

dbus 1.13.16 (2020-06-02)
=========================

The “ominous mushroom hat” release.

Denial of service fixes:

• CVE-2020-12049: If a message contains more file descriptors than can
  be sent, close those that did get through before reporting error.
  Previously, a local attacker could cause the system dbus-daemon (or
  another system service with its own DBusServer) to run out of file
  descriptors, by repeatedly connecting to the server and sending fds that
  would get leaked.
  Thanks to Kevin Backhouse of GitHub Security Lab.
  (dbus#294, GHSL-2020-057; Simon McVittie)

Enhancements:

• The API reference manual can be built as a Qt compiled help file if
  qhelpgenerator(-qt5) is available. This is controlled by
  --enable-qt-help and --with-qchdir in the Autotools build, or
  -DENABLE_QT_HELP and -DINSTALL_QCH_DIR in CMake.
  (dbus!150, Ralf Habacker)

Fixes:

• When built for Windows, return all autolaunch error information in
  the DBusError rather than printing some of it to stderr
  (dbus#191, dbus!131; Ralf Habacker)

• When built for Windows, don't truncate long log messages
  (dbus!134, Ralf Habacker)

• When built using CMake for a Unix platform, dbus-cleanup-sockets and
  dbus-uuidgen are now included (dbus!154, Ralf Habacker)

• When built for Windows with verbose mode enabled, don't print debugging
  messages related to poll() emulation into a fixed-size buffer that
  could overflow (dbus!125, Ralf Habacker)

• Adjust .desktop file parser to avoid a Coverity false positive
  (dbus!146, Coverity CID 354884; Ralf Habacker)

• Print shell-test diagnostics to stderr, avoiding warnings or errors
  from strict TAP parsers (dbus!157, Félix Piédallu)

Tests and CI enhancements:

• When the CI cross-builds Windows binaries on Linux, run unit tests
  using Wine (dbus#296, dbus!158; Ralf Habacker)

• Really build x86_64 Windows binaries in Gitlab-CI, instead of building
  i686 binaries a second time (Ralf Habacker)

• When tests will be run using Wine, use STABS debug symbol format so
  that Wine can display backtraces (dbus#133, dbus!104; Ralf Habacker)

dbus 1.13.14 (2020-04-21)
=========================

The “mystery allium” release.

Dependencies:

• On Unix platforms, if getpwnam_r() and getgrnam_r() are implemented,
  they must be POSIX-conformant. The non-POSIX signature seen in ancient
  Solaris versions will no longer work. (dbus!11, Simon McVittie)

Enhancements:

• D-Bus Specification 0.36:
  · Fix a typo in an annotated hexdump of part of a message
    (dbus!152, Zygmunt Krynicki)

• On Linux, use getrandom(2) in preference to /dev/urandom
  (dbus!147, Natanael Copa)

• Add a --sender option to dbus-send, which requests a name and holds it
  until the signal has been sent. (dbus!116, Christopher Morin)

Fixes:

• Fix a crash when the dbus-daemon is terminated while one or more
  monitors are active (dbus#291, dbus!140; Simon McVittie)

• Fix several test failures if the build-time tests were run as uid 0.
  Note that running the tests with elevated privileges is likely to be
  insecure, and should only be attempted in an expendable container or
  virtual machine. (dbus!117, Simon McVittie)

• Fix an assertion failure if a client encounters an out-of-memory
  condition while sending its response to the "OK" authentication
  message, and processing of the "OK" message is subsequently retried
  when more memory is available (dbus!119, Simon McVittie)

• Don't leak struct addrinfo if we run out of memory during a TCP
  connect()
  (dbus!143, dbus!144, Coverity CID 354880; Ralf Habacker, Simon McVittie)

• On Linux with SELinux, don't assume that the system policy has the
  "dbus" security class or the associated AV
  (dbus#198, dbus!128; Laurent Bigonville)

• Handle dbus_connection_set_change_sigpipe() in a thread-safe way
  (dbus!132; Simon McVittie, Ralf Habacker)

• On Unix, use POSIX <poll.h> in preference to <sys/poll.h>
  (dbus!148, Natanael Copa)

• When building with CMake, cope with libX11 in a non-standard location
  (dbus!129, Tuomo Rinne)

• On Windows with verbose mode enabled and outputting to the debug port,
  use a dynamically-allocated buffer to avoid potential stack buffer
  overflows in long messages (dbus#45, dbus!133; Ralf Habacker)

• The dbus-send(1) man page now documents --bus and --peer instead of
  the old --address synonym for --peer, which has been deprecated since
  the introduction of --bus and --peer in 1.7.6
  (fd.o #48816, dbus!115; Chris Morin)

• Fix a wrong environment variable name in dbus-daemon(1)
  (dbus#275, dbus!122; Mubin, Philip Withnall)

• Fix formatting of dbus_message_append_args example
  (dbus!126, Felipe Franciosi)


dvb-latest/v4l/scripts/rmmod.pl
#!/usr/bin/perl
use strict;
use File::Find;
use Proc::ProcessTable;


my %depend = ();
my %depend2 = ();
my %rmlist = ();
my @nodep;
my @modlist;
my @allmodules;
my %reqmodules;
my %loaded = ();
my $i=0;

# Device debug parameters
#		Module name		   Debug option
my %debug = (	"tuner"			=> "tuner_debug=1",
		"dvb-core"		=> "cam_debug=1",
		"dvb-ttpci"		=> "debug=247",
		"b2c2-flexcop"		=> "debug=0x01",
		"b2c2-flexcop-usb"	=> "debug=0x01",
		"b2c2-flexcop-pci"	=> "debug=0x01",
		"dvb-usb"		=> "debug=0x33",
		"dvb-usb-gp8psk"	=> "debug=0x03",
		"dvb-usb-vp7045"	=> "debug=0x03",
		"dvb-usb-dtt200u"	=> "debug=0x03",
		"dvb-usb-dibusb-common"	=> "debug=0x03",
	    );

sub getobsolete()
{
	my @obsolete;
	open OBSOLETE, '<obsolete.txt' or die "Unable to open obsolete.txt: $!";
	while (<OBSOLETE>) {
		next if (/^\s*#/ || /^\s*$/);
		chomp;
		m|^.*/([^/]+)$| and push @obsolete, $1;
	}

	close OBSOLETE;
	return @obsolete;
}

sub findprog($)
{
	foreach(split(/:/, $ENV{PATH}),qw(/sbin /usr/sbin /usr/local/sbin)) {
		return "$_/$_[0]" if(-x "$_/$_[0]");
	}
	die "Can't find needed utility '$_[0]'";
}

sub parse_dir {
	my $file = $File::Find::name;
	my $modinfo = findprog('modinfo');

	if (!($file =~ /[.]ko$/)) {
		return;
	}

	my $module = $file;
	$module =~ s|^[./]*(.*)[.]ko|\1|;

	open IN, "$modinfo $file|grep depends|cut -b 17-|";
	while (<IN>) {
		my $deps = $_;
		$deps =~ s/\n//;
		$deps =~ s/[,]/ /g;
		$deps = " $deps ";
		$depend{$module} = $deps;
		push @allmodules, $module;
		$i++;
	}
	close IN;
}

sub parse_loaded {
	open IN,  "/proc/modules";
	while (<IN>) {
		m/^([\w\d_-]+)/;
		$loaded{$1}=1;
	}
	close IN;
}

sub cleandep()
{
	my $dep;

	while ( my ($k, $v) = each(%depend) ) {
		my $arg=$v;
		my $arg2=" ";
		while (!($arg =~ m/^\s*$/)) {
			if ($arg =~ m/^ ([^ ]+) /) {
				my $val=$1;
				if (exists($depend{$val})) {
					$arg2="$arg2 $val ";
				} else {
					$reqmodules{$val}=1;
				}
			}
			$arg =~ s/^ [^ ]+//;
			$arg2 =~ s/\s\s+/ /;
		}
		$depend2 { $k } = $arg2;
	}

}

sub rmdep()
{
	my $dep;

	while ($dep=pop @nodep) {
		while ( my ($k, $v) = each(%depend2) ) {
			if ($v =~ m/\s($dep)\s/) {
				$v =~ s/\s${dep}\s/ /;
				$v =~ s/\s${dep}\s/ /;
				$v =~ s/\s${dep}\s/ /;
				$depend2 {$k} = $v;
			}
		}
	}
}

sub orderdep ()
{
	my $old;
	do {
		$old=$i;
		while ( my ($key, $value) = each(%depend2) ) {
			if ($value =~ m/^\s*$/) {
				push @nodep, $key;
				push @modlist, $key;
				$i=$i-1;
				delete $depend2 {$key};
			}
		}
		rmdep();
	} until ($old==$i);
	while ( my ($key, $value) = each(%depend2) ) {
		printf "ERROR: bad dependencies - $key ($value)\n";
	}
}

sub insmod ($)
{
	my $debug=shift;
	my $modprobe = findprog('modprobe');
	my $insmod = findprog('insmod');

	while ( my ($key, $value) = each(%reqmodules) ) {
		print ("$modprobe $key\n");
		system ("$modprobe $key");
	}

	foreach my $key (@modlist) {
		if ($debug) {
			my $dbg=$debug{$key};

			print "$insmod ./$key.ko $dbg\n";
			system "$insmod ./$key.ko $dbg\n";
		} else {
			print "$insmod ./$key.ko\n";
			system "$insmod ./$key.ko\n";
		}
	}
}

my @pulse;
my $try_pulseaudio = 1;

sub check_pulseaudio()
{
	my $t = new Proc::ProcessTable;
	foreach my $p ( @{$t->table} ) {
		push @pulse, $p->uid if ($p->cmndline =~m,/pulseaudio ,);
	}
	$try_pulseaudio = 0 if (!@pulse);

	print "Pulseaudio is running with UUID(s): @pulse\n";
}

sub unload_pulseaudio($)
{
	my $driver_name = shift;
	my $cur_module;

	return if (!$try_pulseaudio);

	check_pulseaudio() if (!@pulse);
	return if (!$try_pulseaudio);

	for my $pid (@pulse) {
		my $user = getpwuid($pid);

#		printf "LANG=C su $user -c \"pacmd list-sources\" |\n";
		open IN, "LANG=C su $user -c \"pacmd list-sources\" |";
		while (<IN>) {
			$cur_module = $1 if (/^\s*module:\s*(\d+)/);

			if (/^\s*alsa.driver_name\s*=\s*"(.*)"/) {
				if ($1 eq $driver_name) {
#					print "LANG=C su $user -c \"pactl unload-module $cur_module\"\n";
					system ("LANG=C su $user -c \"pactl unload-module $cur_module\"");
				}
				next;
			}

			# Special case: em28xx sometimes use a Vendor Class at
			# the same interface as the video node. Pulseaudio can't
			# get the driver name in this case
			if (/^\s*alsa.card_name\s*=\s*"Em28xx/) {
#				print "LANG=C su $user -c \"pactl unload-module $cur_module\"\n";
				system ("LANG=C su $user -c \"pactl unload-module $cur_module\"");
			}
		}
		close IN;

#		printf "LANG=C su $user -c \"pacmd list-sinks\" |\n";
		open IN, "LANG=C su $user -c \"pacmd list-sinks\" |" or return;
		while (<IN>) {
			$cur_module = $1 if (/^\s*module:\s*(\d+)/);
			if (/^\s*alsa.driver_name\s*=\s*"(.*)"/) {
				if ($1 eq $driver_name) {
#					print "LANG=C su $user -c \"pactl unload-module $1\"\n";
					system ("LANG=C su $user -c \"pactl unload-module $1\"");
				}
			}
		}
	}
	close IN;
}

sub rmmod(@)
{
	my $rmmod = findprog('rmmod');
	my @not;
	foreach (reverse @_) {
		s/-/_/g;
		if (exists ($loaded{$_})) {
			my $module = $_;
			print "$rmmod $module\n";
			unload_pulseaudio($module);
			unshift @not, $module if (system "$rmmod $module");
		}
	}
	return @not;
}

sub prepare_cmd()
{
	find(\&parse_dir, ".");
	printf "found $i modules\n";

	cleandep();
	orderdep();
}

# main
my $mode=shift;
if ($mode eq "load") {
		prepare_cmd;
		insmod(0);
} else {
	if ($mode eq "unload") {
		prepare_cmd;
		parse_loaded;
		my @notunloaded = rmmod(@modlist, getobsolete());
		@notunloaded = rmmod(@notunloaded) if (@notunloaded);
		if (@notunloaded) {
			print "Couldn't unload: ", join(' ', @notunloaded), "\n";
		}
	} elsif ($mode eq "reload") {
		prepare_cmd;
		parse_loaded;
		rmmod(@modlist);
		insmod(0);
	} elsif ($mode eq "debug") {
		prepare_cmd;
		parse_loaded;
		insmod(1);
	} elsif ($mode eq "check") {
		prepare_cmd;
		parse_loaded;
	} else {
		printf "Usage: $0 [load|unload|reload|debug|check]\n";
	}
}

